MerlinService#setHostname should update currentNetworkStatusRetriever

joetimmins opened the pull request:
When using `MerlinBuilder#setEndPoint` to build `MerlinService` this field is not used. The problem seems to be that `MerlinService#setHostname` updates `hostPinger` but it doesn't update the `currentNetworkStatusRetriever` with this new `hostPinger`, so it keeps using the default one.

This PR tries to solve this issue by also updating `currentNetworkStatusRetriever`.

joetimmins commented:
This is basically a direct copy of #80 but it's actually mergeable now.

xrigau commented on core/src/test/java/com/novoda/merlin/service/MerlinServiceShould.java:
```diff
@@ -112,6 +117,32 @@ public void callMerlinNetworkListenerOnDisconnectedWhenAConnectedConnectivityEve
+        ContextWrapper contextWrapper = mock(ContextWrapper.class);
+        ConnectivityManager connectivityManager = mock(ConnectivityManager.class);
+        Mockito.when(contextWrapper.getApplicationContext()).thenReturn(context);
+        Mockito.when(context.getSystemService(Context.CONNECTIVITY_SERVICE)).thenReturn(connectivityManager);
```
It'd be nice if this could be a static import 😉  and same for the assertion https://github.com/novoda/merlin/pull/84/files#diff-041de7855cd1408bc7e88f09d0371af3R143

xrigau commented on core/src/test/java/com/novoda/merlin/service/MerlinServiceShould.java:
```diff
@@ -3,19 +3,24 @@
 import org.junit.runners.JUnit4;
 import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.internal.util.reflection.Whitebox;
```
😮  TIL

joetimmins replied:
Pretty much my reaction too! Never seen it before

rock3r replied:
I'm not sure `internal` stuff is meant to be used from outside of Mockito :P

xrigau commented on core/src/test/java/com/novoda/merlin/service/MerlinServiceShould.java:
```diff
@@ -112,6 +117,32 @@ public void callMerlinNetworkListenerOnDisconnectedWhenAConnectedConnectivityEve
+
+        CurrentNetworkStatusRetriever currentNetworkStatusRetriever = (CurrentNetworkStatusRetriever) Whitebox.getInternalState(merlinService, "currentNetworkStatusRetriever");
+        HostPinger hostPinger = (HostPinger) Whitebox.getInternalState(currentNetworkStatusRetriever, "hostPinger");
+        String currentTestHostAddress = (String) Whitebox.getInternalState(hostPinger, "hostAddress");
```
I think this might be a bit of a smell (testing the internals of the class), shouldn't we try to test the _what_ rather than the _how_? (testing what would be the result of resetting the status retriever rather than testing that we reset the retriever)

rock3r replied:
I agree, this seems too much on the whitebox testing side of things (you even use a `Whitebox` helper, that is rather telling)

joetimmins replied:
Yeah I'm completely re-writing this test. This was the test provided by the external contributor which explains why it's using stuff none of us are familiar with.

rock3r commented on core/src/main/java/com/novoda/merlin/service/MerlinService.java:
```diff
@@ -82,6 +82,7 @@ protected ComponentName connectivityReceiverComponent() {
 
     public void setHostname(String hostname, ResponseCodeValidator validator) {
         hostPinger = HostPinger.withCustomEndpointAndValidation(this, hostname, validator);
+        currentNetworkStatusRetriever = new CurrentNetworkStatusRetriever(MerlinsBeard.from(this.getApplicationContext()), hostPinger);
```
You don't need the `this`, right? Besides, this smells a bit of unexpected side effect (but I see the point). Maybe extract a setter for the retriever?

joetimmins commented:
I've added `@VisibleForTesting` factory methods to `MerlinService` so that I can stub out the instantiation of new `HostPinger` objects. So I was able to rewrite the tests to not use `Whitebox`.

I also made sure the `ResponseCodeValidator` could always be set whenever the `hostName` is set on the public API and renamed `CurrentNetworkStatusRetriever` to `NetworkStatusRetriever` because it was bugging me 👅

blundell commented:
visible for testing is a bit of a cop out but it does the job 👍 LGTM

frapontillo commented on core/src/main/java/com/novoda/merlin/service/MerlinService.java:
```diff
@@ -118,7 +118,22 @@ public void onConnectivityChanged(ConnectivityChangeEvent connectivityChangeEven
+    }
+
+    @VisibleForTesting
+    protected HostPinger buildDefaultHostPinger() {
```
Can you move this method closer to where it's actually used for the last time?

joetimmins replied:
Sure thing 👍

joetimmins commented:
Agree it's a cop-out but it's going to be lots of effort to turn that service into a properly testable thing. I tried extracting some of the logic into a pure Java class but it ended up being tiny, almost too tiny to be worth testing.

The other way of doing it would be to use Robolectric but I'm wary of including Robolectric with a project that's a) gone this long without it and b) probably going to be deprecated soon.

frapontillo commented on core/src/main/java/com/novoda/merlin/service/MerlinService.java:
```diff
@@ -118,7 +118,22 @@ public void onConnectivityChanged(ConnectivityChangeEvent connectivityChangeEven
+    }
+
+    @VisibleForTesting
+    protected HostPinger buildHostPinger(String hostName, ResponseCodeValidator validator) {
```
Move this as well please 🔁

frapontillo commented:
I'd move `TestDoubleMerlinService` and `TestDoubleMerlinServiceWithStubbedBuilders` to separate files to improve readability, but not a blocker, up to @joetimmins.

👍 LGTM

joetimmins commented:
Yep good call @frapontillo - will do that now

frapontillo commented:
![shipit](http://i.giphy.com/9NZkSQOyNYLQY.gif)
